db/kvstore_versioned.go

```diff
@@ -47,4 +52,107 @@ type (
 		// SetVersion sets the version, and returns a KVStore to call Put()/Get()
 		SetVersion(uint64) KVStore
 	}
+
+	// KvWithVersion wraps the versioned DB implementation with a certain version
+	KvWithVersion struct {
+		db        VersionedDB
+		kvBase    KVStore
+		versioned map[string]bool // map of versioned buckets
+		version   uint64          // the current version
+	}
 )
+
+// Option sets an option
+type Option func(*KvWithVersion)
+
+func VersionedNamespaceOption(ns ...string) Option {
+	return func(k *KvWithVersion) {
+		k.versioned = make(map[string]bool)
+		for _, ns := range ns {
+			k.versioned[ns] = true
+		}
+	}
+}
+
+// NewKVStoreWithVersion implements a KVStore that can handle both versioned
+// and non-versioned namespace
+func NewKVStoreWithVersion(cfg Config, opts ...Option) *KvWithVersion {
+	db := NewBoltDBVersioned(cfg)
+	kv := KvWithVersion{
+		db:     db,
+		kvBase: db.Base(),
```

Comment:
We may need to review the implementation of BoltDBWithVersion. From the name, `BoltDBWithVersion` sounds like an implementation of `KVStoreWithVersion`. Moreover, `db.Base()` is a weird interface.